src/cache_filesystem.cpp

```diff
@@ -432,17 +499,33 @@ int64_t CacheFileSystem::ReadImpl(FileHandle &handle, void *buffer, int64_t nr_b
 		return 0;
 	}
 
+	auto state = instance_state.lock();
+
 	// If the source filepath matches exclusion regex, skip caching.
 	const int64_t bytes_to_read = MinValue<int64_t>(nr_bytes, file_size - location);
-	if (CacheExclusionManager::GetInstance().MatchAnyExclusion(handle.GetPath())) {
+	auto *exclusion_mgr = GetExclusionManager(state.get());
+	if (exclusion_mgr && exclusion_mgr->MatchAnyExclusion(handle.GetPath())) {
 		auto &cache_handle = handle.Cast<CacheFileSystemHandle>();
 		internal_filesystem->Read(*cache_handle.internal_file_handle, buffer, nr_bytes, location);
 		return bytes_to_read;
 	}
 
 	// Leverage cache read manager.
-	cache_reader_manager.GetCacheReader()->ReadAndCache(handle, static_cast<char *>(buffer), location, bytes_to_read,
-	                                                    file_size);
+	auto *mgr = GetCacheReaderManager(state.get());
+	if (mgr) {
+		auto *reader = mgr->GetCacheReader();
+		if (reader) {
+			reader->ReadAndCache(handle, static_cast<char *>(buffer), location, bytes_to_read, file_size);
+		} else {
+			// Fallback to direct read if no cache reader
+			auto &cache_handle = handle.Cast<CacheFileSystemHandle>();
+			internal_filesystem->Read(*cache_handle.internal_file_handle, buffer, nr_bytes, location);
+		}
+	} else {
```

Comment:
Just want to make sure I understand correct, the else branch here only happens when there's pending IO when database instance destructs. Did you meet such circumstances in your production, or we could simply assertion failure here?

Reply:
We already have a util function `GetInstanceStateOrThrow`, do you think it makes sense to call it directly?